integrations/terraform_cloud/client.py

```diff
@@ -0,0 +1,185 @@
+                params=query_params,
+                json=json_data,
+            )
+            response.raise_for_status()
+
+            logger.info(
+                f"Successfully retrieved Terraform Cloud data for endpoint: {endpoint}"
+            )
+
+            return response.json()
+
+        except httpx.HTTPStatusError as e:
+            logger.error(
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+        except httpx.HTTPError:
+            logger.exception("An error occured while fetching terraform cloud data")
+            raise
+
+    async def get_organizations(self) -> list[dict[str, Any]]:
+        response = await self.send_api_request(endpoint="organizations")
+        return response.get("data", [])
+
+    async def get_single_workspace(self, workspace_id: str) -> dict[str, Any]:
+        response = await self.send_api_request(endpoint=f"workspaces/{workspace_id}")
+        return response.get("data", {})
+
+    async def get_single_run(self, run_id: str) -> dict[str, Any]:
+        response = await self.send_api_request(endpoint=f"runs/{run_id}")
+        return response.get("data", {})
+
+    async def get_state_version_output(self, state_version_id: str) -> dict[str, Any]:
+        response = await self.send_api_request(
+            endpoint=f"state-versions/{state_version_id}/outputs"
+        )
+        return response.get("data", {})
+
+    async def get_paginated_workspaces(
+        self,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        page = 1
+
+        organizations = await self.get_organizations()
+        for organization in organizations:
+            logger.info(f"Getting workspaces for {organization['id']}")
+
+            while True:
+                params: dict[str, Any] = {"page[number]": page, "page[size]": PAGE_SIZE}
+                response = await self.send_api_request(
+                    endpoint=f"organizations/{organization['id']}/workspaces",
+                    query_params=params,
+                )
+                workspaces = response.get("data", [])
+
+                yield workspaces
+
+                if len(workspaces) < PAGE_SIZE:
+                    break
+
+                page += 1
+
+    async def get_paginated_runs_for_workspace(
+        self, workspace_id: str
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info(f"Getting runs for workspace {workspace_id} from Terraform")
+
+        page = 1
+        while True:
+            params: dict[str, Any] = {"page[number]": page, "page[size]": PAGE_SIZE}
+            response = await self.send_api_request(
+                endpoint=f"workspaces/{workspace_id}/runs", query_params=params
+            )
+            runs = response.get("data", [])
+            yield runs
+
+            if len(runs) < PAGE_SIZE:
+                break
+            page += 1
+

```

Comment:
A nice addition, would be to log the metadata returned from their pagination API that tells how many are next and how many we already processed, this will help with possible debugging.
```
  "meta": {
    "pagination": {
      "current-page": 1,
      "prev-page": null,
      "next-page": 2,
      "total-pages": 2,
      "total-count": 21
    }
```